src/inventory_ui.cpp

```diff
@@ -884,9 +892,71 @@ void inventory_column::prepare_paging( const std::string &filter )
             from = to;
         }
     }
+
+    if( !grouped ) {
+        // remove group entries
+        for( const std::shared_ptr< group_entry > &g_iter : group_entries_ ) {
+            // TODO: child entries need to be put back!
+            std::remove( entries.begin(), entries.end(), *g_iter );
+        }
+        group_entries_.clear();
+
+        std::vector<item_location> group;
+        bool inside_group = false;
+        std::vector<inventory_entry>::iterator it, nx, gp;
+        it = nx = gp = entries.begin();
+        while( it != entries.end() ) {
+            nx++;
+            if( nx != entries.end() && it->is_item() && nx->is_item() &&
+                it->any_item()->typeId() == nx->any_item()->typeId() ) {
+                // save items
+                group.insert( group.end(), it->locations.begin(), it->locations.end() );
+                if( !inside_group ) {
+                    gp = it;
+                    inside_group = true;
+                }
+            } else if( inside_group ) {
+                // save items
+                group.insert( group.end(), it->locations.begin(), it->locations.end() );
+                // create group entry
+                int children = std::distance( gp, nx );
+                group_entry g_entry( group, it->get_category_ptr() );
+                g_entry.update_cache();
+                g_entry.highlight_as_parent = true;
+                g_entry.set_number_of_children( children );
+                group_entries_.push_back(  // cant't be in main scope?
+                    std::make_shared<group_entry> ( g_entry ) );
+
+                nx = entries.insert( gp, g_entry );
+                expand_to_fit( *gp );
+
+                std::advance( nx, children + 1 );
+                group.clear();
+                inside_group = false;
+            }
+            it = nx;
+        }
+        grouped = true;
+    }
+
+    for( const std::shared_ptr< group_entry > &g_iter : group_entries_ ) {
+        // FIXME: it's only here for debug, replace with state check
+        volatile bool expd = g_iter->expanded;
+        if( !expd ) { // TODO: change on change
+            // pop group
+            std::vector<inventory_entry>::iterator gp =
+                std::find( entries.begin(), entries.end(), *g_iter );
+            gp++;
+            entries.erase( gp, gp + g_iter->get_number_of_children() );
+            // TODO: not yet, implement proper state maitanence for `group_entry` first
+            // g_iter->expanded = false;
+        }
+    }
```

Comment:
This caused so many SIGSEGV, that I'm still afraid somehow I messed up it.

Reply:
You might want to sanity-check the variables and print debug messages / bail out early to avoid hard crashes. (For example, check that `g_iter` is not null, `gp` is not `entries.end()`, `std::distance( gp, entries.end() )` is not lower than `g_iter->get_number_of_children()`